pkgs/native_assets_cli/lib/src/config.dart

```diff
@@ -426,12 +483,58 @@ class BuildOutputBuilder extends HookOutputBuilder {
       syntax.BuildOutput.fromJson(super._syntax.json);
 }
 
+extension type MetadataOutputBuilder._(BuildOutputBuilder _output) {
+  void operator []=(String key, Object value) {
+    _output.assets.addEncodedAsset(
+      MetadataAsset(key: key, value: value).encode(),
+      routing: const ToBuildHooks(),
+    );
+  }
+
+  void addAll(Map<String, Object> metadata) {
+    for (final MapEntry(:key, :value) in metadata.entries) {
+      this[key] = value;
+    }
+  }
+}
+
+/// The destination for assets output in a build hook.
+sealed class AssetRouting {
+  const AssetRouting();
+}
+
+/// Assets with this routing will be bundled in the final application.
+final class BundleInApp extends AssetRouting {
+  const BundleInApp();
+}
+
+/// Assets will be send to build hooks.
+///
+/// The assets are only available for build hooks of packages that have a direct
+/// dependency on the package emitting the asset with this routing.
+///
+/// The assets will not be bundled in the final application unless added with
+/// [BundleInApp] in another hook.
+final class ToBuildHooks extends AssetRouting {
+  const ToBuildHooks();
+}
+
+/// Assets will be send to the link hook of [packageName].
+///
+/// The assets are only available to the link hook of [packageName].
+///
+/// The assets will not be bundled in the final application unless added with
+/// [BundleInApp] in the link hook of [packageName].
+final class ToLinker extends AssetRouting {
+  final String packageName;
+
+  const ToLinker(this.packageName);
```

Comment:
Nit: Maybe `LinkIn`? I like it when method names and the parameter make a sort of sentence.